sql-plugin/src/main/scala/com/nvidia/spark/rapids/Arm.scala

```diff
@@ -156,6 +158,34 @@ trait Arm {
       h.close()
     }
   }
+
+  def withRetry(
```

Comment:
nit: I don't like these being in `Arm`. Almost all of the other implementations here are about AutoClosable. That makes the dependencies very simple.

There is one for freeing a `RapidsBuffer`, which that is kind of okay because `RapidsBuffer` is a trait, but it is defined in a file where Arm is used. This means the files have a lot of coupling between each other. 

But `RmmRapidsRetryHelper` extends `Arm`.  The two are mutually dependent and I really don't like that. Yes the tools support it, but from a design perspective this is not great.  Can we have a separate implicit include class for this? And possibly a follow on to do something similar with freeOnExcept?

Reply:
Should the end result be an implicit? Or should it be a trait like `Arm`?

implicit: `mySourceIter.withRetry { doWork(_) }`
trait:  `withRetry(mySourceIter) { doWork(_) }`

The main advantage with the implicit, is we don't need to extend a trait, like with Arm..

Reply:
Can we use a static method that we can simply import directly?  Even if it's an implicit, we'd have to import the implicit declaration to use it, so I'm wondering if we can make this more straightforward by simply importing the static methods directly instead.

Reply:
Ah, that sounds good to me.

Reply:
Went the static route, they are defined in an awkward object `RmmRapidsRetryIterator`, but can create something else like `RmmRapidsRetry`.